### PR TITLE
#38: speed-up chain loading by caching blocks and hashes

### PR DESCRIPTION
**Pull request by atramos**
I ran a profiler and found most of the time loading the chain was spent on disk and sha256 hash operations.
With this PR, chain loads approximately 20x faster on a cold-started machine, 3x faster on a pre-warmed machine.

### CODE REVIEW

**Comment by lk16** [comment]
Nice, I was just looking into making the chain load faster as well. However I took a completely [different](https://github.com/lk16/skepticoin/blob/compressed-chain/skepticoin/files/chain.py) approach, because I didn't run a profiler like you. I assumed it was slow due to loading so many small files and tried to put them together in a tar.gz. 

Above link is a work in progress by the way, it's a big mess.

Regarding your PR:
Caching the hashvalue in `Transaction` makes a lot of sense.

I'm not a fan of dumping and loading pickled data. Especially considering the big security risks it involves. 
Why not do one of these:
* Store the data you pickle right now but just as a JSON file. (Possibly adding more data from CoinState)
* Change the block files into a JSON, having the block data be a base64 encoded string, together with the block's hash. 

> 308 seconds for loading chain of 73000

Are you using a HDD? 
Can you try out my branch as linked above?

On my branch it takes 27 seconds to load 73000 blocks.
I did not try yet if it becomes faster without compression.

(I'm sorry for talking so much about my own hacking here. I'll open my own PR later when I'm happy with my own code.)

**Comment by atramos** [comment]
I have SSDs. 

Python json doesn't play well with the immutable maps that are used throughout the code. 

The blockchain is public so there is no security issue with using pickle here. 

My PR is not perfect but it's working now and gives a huge developer productivity boost, by not having to wait so long each time, so I hope we can get it merged soon so that I can move on to other needed changes.

File: skepticoin/datatypes.py
Comment (sashimi-houdini) [inline skepticoin/datatypes.py:132]: This approach hinges on the immutability of transactions. In practice I think I've been smart enough to never treat transactions as mutable objects, but I know for a fact the immutability is never enforced. Since this particular line really creates a hard dependency on immutability, it's probably for the best to at least note that explicitly in a comment or even in the Transaction's docstring.

File: skepticoin/datatypes.py
Comment (sashimi-houdini) [inline skepticoin/datatypes.py:275]: Same as above

File: skepticoin/scripts/utils.py
Comment (sashimi-houdini) [inline skepticoin/scripts/utils.py:83]: Bikeshedding: `snake_case`

File: skepticoin/coinstate.py
Comment (sashimi-houdini) [inline skepticoin/coinstate.py:133]: Let's just leave this out of the code then

**Review by sashimi-houdini — CHANGES_REQUESTED** [review]
By addressing the small points here I think this will just make it in; I'll make broader comments on the related issue, but those shouldn't hold up the practical gains that this PR brings.

File: skepticoin/datatypes.py
Comment (lk16) [reply skepticoin/datatypes.py:132]: Maybe mypy help with marking this variable as immutable?

File: skepticoin/scripts/utils.py
Comment (atramos) [reply skepticoin/scripts/utils.py]: d7da1212348a9770da347bc300001647bd8f3b60

File: skepticoin/datatypes.py
Comment (atramos) [reply skepticoin/datatypes.py:138]: mypy helps ... 07b92324b80fe36effa4c69df1cfc571a82e5c12

File: skepticoin/datatypes.py
Comment (atramos) [reply skepticoin/datatypes.py:281]: d4f5d3282e9ec4274c7283843281b24b6369de49

**Comment by lk16** [comment]
You have some failing tests and other issues.

`typing.Final` is [new](https://docs.python.org/3/library/typing.html#typing.Final) in python 3.8.
@sashimi-houdini do we _want_ to keep running python 3.7? Otherwise, I'm not sure how to get out of this one.

Consider setting up pre-commit hooks as described [here](https://github.com/skepticoin/skepticoin/blob/master/CONTRIBUTING.md) so you don't have to wait for approval to run CI on your PR. (We need to approve manually until at least one of your MR's gets merged.)

**Comment by atramos** [comment]
My local hooks don't seem to be running. I guess the easy answer here is to remove the Final for now until Python 3.7 reaches EOL.

**Comment by atramos** [comment]
> `typing.Final` is [new](https://docs.python.org/3/library/typing.html#typing.Final) in python 3.8.

I got rid of typing.Final in cc842caa11718f5edd95f9f900f7882b6ec5f402

**Comment by atramos** [comment]
@sashimi-houdini @lk16 please re-review

File: skepticoin/datatypes.py
Comment (lk16) [inline skepticoin/datatypes.py:132]: These would be type annotations for fields of the class Transaction, not instances of it.

Fields of class instances usually get the right type (from mypy) in `__init__()`, either from default constants by assigning  from typed arguments. 

Please remove these lines.

File: skepticoin/datatypes.py
Comment (lk16) [inline skepticoin/datatypes.py:278]: Same as above.

File: tests/test_utils.py
Comment (lk16) [inline tests/test_utils.py:1]: Please remove this file. We don't test any code correctness and profiling can be done without this file.

**Review by lk16 — CHANGES_REQUESTED** [review]
I missed some things, sorry about that. Please have a look

File: skepticoin/datatypes.py
Comment (atramos) [reply skepticoin/datatypes.py:132]: fixed @ f594977

File: skepticoin/datatypes.py
Comment (atramos) [reply skepticoin/datatypes.py:278]: fixed @ f594977

File: tests/test_utils.py
Comment (atramos) [reply tests/test_utils.py]: Moved it to `performance_tests/`. Is that ok? I think performance testing is important.
commit 2c7ac1a

Comment (lk16) [reply tests/test_utils.py]: Works for me. It's clear they're not regular tests and they won't be run by CI or pre-commit hooks.

**Comment by atramos** [comment]
Finally all checks passed!

**Comment by lk16** [comment]
Very nice. I'll merge this one, since all comments from @sashimi-houdini are addressed as far as I can tell.

**Comment by atramos** [comment]
@lk16 I'm going to reopen this and cherry-pick a commit out of the cloud mining branch, which is required for `feature/fast-chainload` and `feature/fresh-balance` to work correctly together.

**Comment by atramos** [comment]
Actually no, I'll open a separate PR.

**Pull request merged**
